Source/JavaScriptCore/runtime/JSArrayBufferPrototype.cpp

```diff
@@ -293,10 +295,75 @@ JSC_DEFINE_HOST_FUNCTION(arrayBufferProtoFuncResize, (JSGlobalObject* globalObje
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    ASSERT(arrayBuffer->impl()->sharingMode() == ArrayBufferSharingMode::Default);
+    bool isResizable = arrayBuffer->isResizableOrGrowableShared();
+
+    if (UNLIKELY(arrayBuffer->impl()->isDetached())) {
+        throwVMTypeError(globalObject, scope, "Receiver is detached"_s);
+        return nullptr;
+    }
+
+    if (!isResizable && newByteLength == arrayBuffer->impl()->byteLength()) {
+        // We should just transfer!
+        ArrayBufferContents contents;
+        if (UNLIKELY(!arrayBuffer->impl()->transferTo(vm, contents))) {
+            throwVMRangeError(globalObject, scope, makeString("transfer failed"_s));
+            return nullptr;
+        }
+        auto newBuffer = ArrayBuffer::create(WTFMove(contents));
+        return JSArrayBuffer::create(vm, globalObject->arrayBufferStructure(ArrayBufferSharingMode::Default), WTFMove(newBuffer));
+    }
+
+    if (mode == CopyAndDetachMode::PreserveResizability && isResizable) {
+        if (UNLIKELY(newByteLength > arrayBuffer->impl()->maxByteLength())) {
+            throwVMRangeError(globalObject, scope, makeString("transfer failed with new byte length "_s, newByteLength));
+            return nullptr;
+        }
+
+        ArrayBufferContents contents;
+        if (UNLIKELY(!arrayBuffer->impl()->transferTo(vm, contents))) {
+            throwVMRangeError(globalObject, scope, makeString("transfer failed"_s));
+            return nullptr;
+        }
+        auto newBuffer = ArrayBuffer::create(WTFMove(contents));
+        if (!newBuffer->resize(vm, newByteLength)) {
+            throwVMRangeError(globalObject, scope, makeString("resize failed with new byte length "_s, newByteLength));
+            return nullptr;
+        }
+        return JSArrayBuffer::create(vm, globalObject->arrayBufferStructure(ArrayBufferSharingMode::Default), WTFMove(newBuffer));
+    }
 
+    // We should create a new ArrayBuffer and copy them since underlying ArrayBuffer characteristics are different.
+    auto newBuffer = ArrayBuffer::tryCreate(newByteLength, 1, std::nullopt);
+    if (UNLIKELY(!newBuffer)) {
+        throwOutOfMemoryError(globalObject, scope);
+        return nullptr;
+    }
+    size_t copyLength = std::min<size_t>(newByteLength, arrayBuffer->impl()->byteLength());
+    memcpy(newBuffer->data(), arrayBuffer->impl()->data(), copyLength);
+
+    ArrayBufferContents dummyContents;
+    if (UNLIKELY(!arrayBuffer->impl()->transferTo(vm, dummyContents))) {
+        throwVMRangeError(globalObject, scope, makeString("transfer failed"_s));
+        return nullptr;
+    }
+
+    return JSArrayBuffer::create(vm, globalObject->arrayBufferStructure(ArrayBufferSharingMode::Default), WTFMove(newBuffer));
+}
+
+// https://tc39.es/proposal-arraybuffer-transfer/#sec-arraybuffer.prototype.transfer
+JSC_DEFINE_HOST_FUNCTION(arrayBufferProtoFuncTransfer, (JSGlobalObject* globalObject, CallFrame* callFrame))
```

Comment:
Since `arrayBufferProtoFuncTransfer` and `arrayBufferProtoFuncTransferToFixedLength` are literally identical except that they pass a different `CopyAndDetachMode` value to `arrayBufferCopyAndDetach` at the end, can you please create a common static function that both call with the `CopyAndDetachMode` as an argument?  Let's not copy paste code like this unnecessarily.  

In fact, it looks like the spec itself is doing this i.e. sharing code except for just the `CopyAndDetachMode` argument.  So, it looks like you can move most of this common functionality into `arrayBufferCopyAndDetach`.

Reply:
I deliberately did this since I don't want to put these kind of argument handlings into `arrayBufferCopyAndDetach` and I would like to eventually move this to `ArrayBuffer::copyAndDetach`, which will take these arguments. I would like to keep it as is, intentionally.

Reply:
In that case, can't you have a `arrayBufferProtoFuncTransferImpl` that calls `arrayBufferCopyAndDetach` instead?  Have `arrayBufferProtoFuncTransfer` and arrayBufferProtoFuncTransferToFixedLength` call `arrayBufferProtoFuncTransferImpl` with a different `CopyAndDetachMode `.

Reply:
Hmm, TBH, I don't agree that this makes code better..., or rather, it adds unnecessary complicated abstraction. But if you strongly think, I can just change.